src/lib.rs

```diff
@@ -10,20 +10,51 @@ pub struct Service {
     tree: Tree<Memory>,
 }
 
+fn valid_namespace(namespace: &[&str]) -> bool {
+    namespace.len() >= 2
+        && namespace.iter().all(|part| {
+            !part.is_empty()
+                && part
+                    .find(|c| !matches!(c, '0'..='9' | 'a'..='z' | 'A'..='Z' | '-'))
+                    .is_none()
+        })
+}
+
 #[async_trait]
 impl Handler<()> for Service {
-    type Response = Response;
+    type Response = Result<Response>;
 
     async fn handle(self, mut req: Request) -> Self::Response {
+        fn no_route() -> Error {
+            Error::from_str(StatusCode::NotFound, "Route not found")
+        }
+
         let url = req.url_mut();
-        let path = url.path().trim_start_matches('/');
-
-        if path == "_tree" || path.starts_with("_tree/") {
-            let path = format!("/{}", &path[5..].trim_start_matches('/'));
-            url.set_path(&path);
-            self.tree.handle(req).await
-        } else {
-            StatusCode::MethodNotAllowed.into()
+        let path = url.path();
+        let (namespace, path) = path
+            .strip_prefix('/')
+            .expect("invalid URI")
+            .split_once("/_")
+            .ok_or_else(no_route)?;
+        let namespace = namespace.split('/').collect::<Vec<&str>>();
+        if !valid_namespace(&namespace) {
+            return Err(Error::from_str(
+                StatusCode::BadRequest,
+                format!("Invalid namespace `{}`", namespace.join("/")),
+            ));
+        }
+
+        let path = path.to_string();
+        let (comp, path) = path.split_once('/').unwrap_or((&path, ""));
+        url.set_path(&format!("/{}", path));
+
+        // TODO: use `namespace`
+
+        match comp {
+            "tree" => Ok(self.tree.handle(req).await),
+            // TODO: use tag service
+            //"tag" => Ok(self.tag.handle(req).await),
+            _ => Err(no_route()),
```

Comment:
I think that namespace should be a type, similar to `path::Path` in the `tree` crate.

Reply:
Did that, focus of this PR the parsing rather than namespace model itself.
It is still not clear how the namespace will actually be used by the components, therefore I left all fields private by design.

We will probably want to implement some conversion traits on the type for usage in components in the future, but requirements are not clear yet.